### ISSUE TITLE
Chaining

### ISSUE THREAD

**Issue by jeff-hykin** [jeff-hykin/vscode-textmate#2]
This feature might be able to solve many of the unsolvable problems and significantly improve performance.

```ruby
PatternRange.new(
    repository_name: "parameters",
    start_pattern: /\(/,
    end_pattern: /\)/,
)
newPattern(
    repository_name: "function_specifier"
    require_previous_match: [ "parameters" ],
    match: std_space.then(/const|mutable|noexcept|etc/),
)
PatternRange.new(
    require_previous_match: [ "parameters", "function_specifier" ]
    start_pattern: /\{/,
    end_pattern: /\}/
)
```
If there was a `prev` var that kept track of the name of the last-matched pattern, and then a simple check that would disqualify patterns based on that, thats all that would be needed to implement the feature.

**Comment by jeff-hykin** [comment]
We're basically already attempting to do the `require_previous_match` via lookbehinds, which is effectively double-computation and is imperfect.

**Comment by matter123** [comment]
When would the previous pattern no longer be valid?

Would `(...) new-cpp2a-specifier const {};` be valid?

**Comment by jeff-hykin** [comment]
I was thinking that no: `var` would be assigned to null after even 1 non-matched character. 

This would prevent all of the performance issues we have with `std_space`. `std_space` would be a standalone/root pattern, then almost all other patterns would have it as a pre-requirement before they could match. This way std_space would only be matched 1 time

**Comment by matter123** [comment]
So correctly it would look like?
```
PatternRange.new(
    require_previous_match: [ "parameters", "std_space", "function_specifier" "std_space"]
    start_pattern: /\{/,
    end_pattern: /\}/
)
```

**Comment by jeff-hykin** [comment]
Oh I actually meant for it to be an "or" statement, but your design could be better. There's somewhat of an issue with optional patterns like "function_specifier"

**Comment by matter123** [comment]
so `require_previous_match: ["parameters", "std_space?", "function_specifier?", "std_space?"]`?

**Comment by jeff-hykin** [comment]
Yeah that could be a good design. `prev` would need to be an array, but it could be compared in an efficient amount of time. It almost certainly has to be more efficient than look-behinds.

**Comment by jeff-hykin** [comment]
The alternative design (not a list) would have been
```
PatternRange.new(
    repository_name: "parameters",
    start_pattern: /\(/,
    end_pattern: /\)/,
)
PatternRange.new(
    repository_name: "space_after_parameters",
    require_previous_match:  "parameters" ,
    start_pattern: lookBehindFor( /\)/ ),
    end_pattern: lookAheadFor( /[^\s]/ ),
    includes: [
        :block_comment
    ] 
)
newPattern(
    repository_name: "function_specifier"
    require_previous_match:  "space_after_parameters" ,
    match: std_space.then(/const|mutable|noexcept|etc/),
)
PatternRange.new(
    repository_name: "space_after_function_specifier",
    require_previous_match:  "function_specifier" ,
    start_pattern: lookBehindFor( /\)/ ),
    end_pattern: lookAheadFor( /[^\s]/ ),
    includes: [
        :block_comment
    ] 
)
PatternRange.new(
    require_previous_match: "space_after_function_specifier" or "space_after_parameters",
    start_pattern: /\{/,
    end_pattern: /\}/
)
```

But its pretty bloated to write. At minimum there would've been ruby-lib code to minimize it

**Comment by matter123** [comment]
And popping a rule of would clear (or restore to the prior point) prev?

**Comment by jeff-hykin** [comment]
I think the popping/pushing behavior would stay the same. (I'm not sure if thats a "yes" or "no" to your question)
```
initial_context = [
    :all_the_normal_initial_context_stuff,

   :parameters,
   :space_after_parameters,
   :function_specifier,
   :space_after_function_specifier,
   :function_body,
]
```
It would be a little messy since they're conceptually a chain, but internally they're all dumped in the same scope and just disqualified most of the time. But on the designer side we could do:
```
inital_context = [
    :all_the_normal_stuff,
    :function_definition
]
grammar[:function_definition] = [
   :parameters,
   :space_after_parameters,
   :function_specifier,
   :space_after_function_specifier,
   :function_body,
]
```

**Comment by jeff-hykin** [comment]
This could also be done a bit different/safer but a bit harder to implement.

```
PatternRange.new(
    repository_name: "parameters",
    start_pattern: /\(/,
    end_pattern: /\)/,
    valid_next_patterns: [
        :space_after_parameters
    ]
)
```

There could be a `following_scopes` array separate from the main pattern stack. If the array was empty, then things behave as usual. If there were patterns in the `following_scopes` array, then the array would be used *instead of* the normal pattern stack.

As soon as a pattern is found/confirmed, if it has a `valid_next_patterns` then the `following_scopes` is *overwritten* with those patterns. If the `valid_next_patterns` doesn't exist, then the `following_scopes` is set back to an empty array. (which would effectually return back to the original context before the chaining started)

**Comment by matter123** [comment]
valid_next_scopes does not affect the validity of the scope that set it, correct?

**Comment by jeff-hykin** [comment]
yeah, it would only be for what was matched next. It wouldn't mutate any existing scopes/stacks

**Comment by matter123** [comment]
Ok so that feature can actually be implemented now.
with roughly:
```json
{
"begin": "some pattern here",
"end": "\\G\\s*",
"patterns": [
	{"include": "#valid_next"}
]
}
```

**Comment by jeff-hykin** [comment]
I think `patterns` is only for whats inside of the range, the `valid_next_patterns` would be for what comes after it

**Comment by jeff-hykin** [comment]
It would leave the parent patterns alone (it wouldn't mutate them) but in the searching process it would override/sidestep them entirely when `valid_next_patterns` were present

**Comment by matter123** [comment]
yeah the range is from the end of "some_pattern_here" (presumably from `newPattern(:match)` and the end is the moment a pattern is captured

**Comment by jeff-hykin** [comment]
So could you do something like? (ignoring space/specifiers for convenience here)
```ruby
PatternRange.new(
    start_pattern: /\(/,
    end_pattern: /\)/,
    valid_next_patterns: [
        :function_body
    ]
)
grammar[:function_body] = PatternRange.new(
    start_pattern: /\{/,
    end_pattern: /\}/,
)
```

**Comment by matter123** [comment]
Maybe. The standard translation involves turning a single match into a range, but turning a range into multiple ranges might work.

**Comment by jeff-hykin** [comment]
okay 👍 let me know if you find anything

**Comment by matter123** [comment]
Best that I can do is by chaining begin/ends with look aheads.
Example grammar. https://gist.github.com/matter123/8fffc7225140bf742327f135e37de50d

parses like:
![Screenshot from 2019-07-06 12-16-30](https://user-images.githubusercontent.com/714007/60760286-ef25bb80-9fe7-11e9-88db-c397bf0f304a.png)

```
next valid
check for next valid

begin
next
end next valid
```

Should be useful https://github.com/microsoft/vscode-textmate/issues/41

**Comment by jeff-hykin** [comment]
That `L:` behavior sounds interesting, I'll have to explore it sometime.
